src/remote/remote.go

```diff
@@ -0,0 +1,326 @@
+// Package remote provides our interface to the Google remote execution APIs
+// (https://github.com/bazelbuild/remote-apis) which Please can use to distribute
+// work to remote servers.
+package remote
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/bazelbuild/remote-apis/build/bazel/semver"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	bs "google.golang.org/genproto/googleapis/bytestream"
+	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/encoding/gzip" // Registers the gzip compressor at init
+	"gopkg.in/op/go-logging.v1"
+
+	"github.com/thought-machine/please/src/core"
+	"github.com/thought-machine/please/src/fs"
+)
+
+var log = logging.MustGetLogger("remote")
+
+// Timeout to initially contact the server.
+const dialTimeout = 5 * time.Second
+
+// Timeout for actual requests
+const reqTimeout = 2 * time.Minute
+
+// Maximum number of times we retry a request.
+const maxRetries = 3
+
+// The API version we support.
+var apiVersion = semver.SemVer{Major: 2}
+
+// A Client is the interface to the remote API.
+//
+// It provides a higher-level interface over the specific RPCs available.
+type Client struct {
+	actionCacheClient pb.ActionCacheClient
+	storageClient     pb.ContentAddressableStorageClient
+	bsClient          bs.ByteStreamClient
+	initOnce          sync.Once
+	state             *core.BuildState
+	err               error // for initialisation
```

Comment:
Do we want to check this in method calls and break early?

Reply:
Ah damn I fixed that but must be on another branch. Yes, good spot :)